client/src/routes/FightSocket.ts

```diff
@@ -0,0 +1,112 @@
+};
+
+namespace FightSocket {
+    export type Listener = (...args: any[]) => void;
+};
+
+export class FightSocket  {
+    private socket: Socket<DefaultEventsMap, DefaultEventsMap>;
+    #newEvents: Writable<Response.NewEvent>;
+    #redSuggestion: Writable<Response.Suggestion>;
+    #blueSuggestion: Writable<Response.Suggestion>;
+
+    constructor(private fightId: string, private judgeId: string) {
+        this.socket = io(location.host);
+
+        this.socket.on('connect', () => console.log('Connection established.'));
+        this.socket.on('disconnect', console.warn);
+        this.socket.on('connect_error', console.error);
+        this.#newEvents = writable();
+    }
+
+    awaitNewEvents() {
+        if (!this.socket.hasListeners(Events.NewEvents)) {
+            this.socket.on(Events.NewEvents, (response: Response.NewEvent) => {
+                this.#newEvents.set(response);
+            });
+        }
+
+        return this.#newEvents;
+    }
+
+    awaitSuggestions() {
+        const listener = (response: Response.Suggestion) => {
+            switch (response.judgeColor) {
+                case "RED":
+                    return this.#redSuggestion.set(response);
+
+                case "BLUE":
+                    return this.#blueSuggestion.set(response);
+            }
+        };
+
+        if (this.socket.hasListeners(Events.EventsSuggestion))
+            this.socket.on(Events.EventsSuggestion, listener);
+
+        return [this.#redSuggestion, this.#blueSuggestion];
+    }
+
+    get newEvents() {
+        return this.#newEvents;
+    }
+
+    private promiseFor<T extends Response.Base>(event: Events): Promise<T> {
+        return new Promise((resolve, reject) => {
+            this.socket.once(event, (response: T) => {
+                response.status !== 'OK'
+                    ? reject(response.status)
+                    : resolve(response);
+            });
+        });
+    }
+
+    join() {
+        this.socket.emit(Events.Join, this.getIds());
+        return this.promiseFor<Response.Join>(Events.Join);
+    }
+
+    startFight() {
+        // TODO: emit only if we are mainJudge (when we have a way to know)
+        this.socket.emit(Events.StartFight, this.getIds());
+        return this.promiseFor<Response.Base>(Events.StartFight);
+    }
+
+    finishFight() {
+        // TODO: same as startFight
+        this.socket.emit(Events.FinishFight, this.getIds());
+        return this.promiseFor<Response.Base>(Events.FinishFight);
+    }
+
+    pauseTimer() {
+        this.socket.emit(Events.PauseTimer, this.getIds());
+    }
+
+    resumeTimer() {
+        this.socket.emit(Events.ResumeTimer, this.getIds());
+    }
+
+    /// Considering if it is useful
+    on(event: Events, listener: FightSocket.Listener) {
+        this.socket.on(event, listener);
+    }
+
+    private getIds() {
+        return { fightId: this.fightId, judgeId: this.judgeId };
+    }
+}
```

Comment:
Nie widzę tutaj nasłuchiwania na warunek zakończenia pojedynku. Brakuje tego czy to będzie do zrobienia w późniejszym czasie?

Reply:
Myślałem, że to będzie do zrobienia w późniejszym czasie, ale mogę to zaimplementować

Reply:
Spoko, może być teraz lub później, chodzi mi po prostu żeby sobie o tym nie zapomnieć.